### ISSUE TITLE
Escape key to end current road

### ISSUE THREAD

**Issue by toozinger** [citybound/citybound#141]
It would be massively convenient to end the current road with the esc key, instead of having to double-click the control point.

Thanks!

-Toozinger

**Comment by aeickhoff** [comment]
Good idea!

**Comment by nimper** [comment]
I think the "Escape" key should cancel the input instead of relying on ctrl-Z to undo it.

Double clicking to end the road seems fine. You click to start, double click to finish. Makes sense to me.

**Comment by chances** [comment]
@nimper So, if I'm drawing a road and moving the cursor around previewing a new stroke, pressing Escape would cancel the current previewed stroke and finish the path I was drawing at the last placed control point?

I like that idea. :+1:

**Comment by nimper** [comment]
Yes, I like that idea too. If you're three or four control points deep you maybe don't want to start from the beginning!

**Comment by hubble618** [comment]
So it seems like there are different options people may want, cancel the most recent stroke, or cancel the full stroke, or something else. Perhaps the best option would be to have it user assignable.

**Comment by chances** [comment]
@hubble618 Do you mean this? An option between:

- Canceling the current unplaced stroke (which could be part of a larger path or simply a single unplaced stroke), or
- Cancelling the entire path as a whole

**Comment by chances** [comment]
Maybe pressing escape once cancels the previewed stroke and finished the path (or removes it in the case of a singular stroke). And then pressing escape again could cancel the current plan, removing it from the canvas. All the while preserving these intermediate changes on the undo stack?

Or maybe the delete key could cancel the plan if nothing is selected? (Maybe I'm getting out of scope :stuck_out_tongue_closed_eyes:)

**Comment by nimper** [comment]
Whatever the solution, I think it should be as intuitive and context-aware as possible. I don't like the idea of hunting through option menus to enable this or that functionality.

**Comment by chances** [comment]
@nimper Yes, **context-aware** FTW!

**Comment by graup** [comment]
To me, this would be the most intuitive:

- ESC to cancel the current previewed stroke and finish the path I was drawing at the last placed control point. (This is similar to how drawing paths in Photoshop or OpenStreetMap works.)
- While drawing, CMD+Z to remove the last control point, keep drawing from the point before that
- After finishing the path, CMD+Z would remove the whole path I last drew

**Comment by mallenwho** [comment]
To ad to ^^ this, I feel that right clicking is a lot more natural way to cancel an in-progress path, keeping whatever other completed nodes is. Saves touching the keyboard. I'd make it escape too.

But that's the bonus of user rebindable keystrokes eh?

**Comment by cauebs** [comment]
Double clicking is not something pleasant. Why not do it like most other games, which require you to click and hold to create roads? That way, when you want to stop drawing, you just release the button. Esc could be used to cancel the current plan. What do you guys think?
